File: src/hyperLiquid_ts_deprecated/functions/gatekeeper.ts
import { APIGatewayProxyEvent, APIGatewayProxyResult } from "aws-lambda";
import { webhookSchema } from "../validators/webhookSchema";
import { WebhookPayload } from "../models/webhook";
import { HTTP } from "../constants/http";
import { SQSClient, SendMessageCommand } from "@aws-sdk/client-sqs";

const sqs = new SQSClient({ region: process.env.AWS_REGION || "us-east-1" });

/**
 * Gatekeeper Lambda
 * Fast validation and immediate push to SQS. Target < 200ms handshake.
 */
export const handler = async (event: APIGatewayProxyEvent): Promise<APIGatewayProxyResult> => {
    try {
        if (!event.body) {
            return { statusCode: 400, body: JSON.stringify({ message: "No body provided" }) };
        }
        
        const body = JSON.parse(event.body);

        /** 
        * Step 1: Validate schema (Fast Joi validation)
        */
        const rawPayload = await webhookSchema.validateAsync(body, { abortEarly: false }) as WebhookPayload;

        /**
         * Step 2: Push to Queue for background processing (Instant SQS Push)
         */
        const queueUrl = process.env.SQS_QUEUE_URL;
        if (!queueUrl) throw new Error("SQS_QUEUE_URL not configured");

        await sqs.send(new SendMessageCommand({
            QueueUrl: queueUrl,
            MessageBody: JSON.stringify(rawPayload)
        }));

        /**
         * Step 3: Return success to TradingView (Prevents timeouts)
         */
        return {
            statusCode: HTTP.OK,
            body: JSON.stringify({ message: "Signal received and queued successfully." })
        };

    } catch (error: any) {
        console.error("Gatekeeper Error:", error);
        
        // Return 400 for validation errors or 500 for critical failures
        const isValidationError = error.isJoi || error.name === 'ValidationError';
        return {
            statusCode: isValidationError ? HTTP.BAD_REQUEST : 500,
            body: JSON.stringify({ error: error.message || "Internal Server Error" })
        };
    }
}

File: src/hyperliquid_python/functions/gatekeeper.py
import json
import os
import boto3
from pydantic import ValidationError
from models.webhook import WebhookPayload
from constants.http import HTTP

sqs = boto3.client('sqs', region_name=os.environ.get("AWS_REGION", "us-east-1"))

def handler(event, context):
    try:
        body_str = event.get('body')
        if not body_str:
            return {
                "statusCode": HTTP.BAD_REQUEST,
                "body": json.dumps({"message": "No body provided"})
            }
            
        body = json.loads(body_str)
        
        # Step 1: Validate schema (Fast Pydantic validation)
        raw_payload = WebhookPayload(**body)
        
        # Step 2: Push to Queue for background processing
        queue_url = os.environ.get("SQS_QUEUE_URL")
        if not queue_url:
            raise ValueError("SQS_QUEUE_URL not configured")
            
        sqs.send_message(
            QueueUrl=queue_url,
            MessageBody=raw_payload.model_dump_json()
        )
        
        # Step 3: Return success to TradingView
        return {
            "statusCode": HTTP.OK,
            "body": json.dumps({"message": "Signal received and queued successfully."})
        }
        
    except ValidationError as e:
        print("Gatekeeper Validation Error:", e)
        return {
            "statusCode": HTTP.BAD_REQUEST,
            "body": json.dumps({"error": e.errors()})
        }
    except Exception as e:
        print("Gatekeeper Error:", e)
        return {
            "statusCode": HTTP.INTERNAL_SERVER_ERROR,
            "body": json.dumps({"error": str(e) or "Internal Server Error"})
        }
